File: src/StarkEx/MessageSigner.cpp
// SPDX-License-Identifier: Apache-2.0
//
// Copyright © 2017 Trust Wallet.

#include <StarkEx/MessageSigner.h>
#include <HexCoding.h>

namespace TW::StarkEx {

std::string MessageSigner::signMessage(const TW::PrivateKey& privateKey, const std::string& message) {
    auto digest = parse_hex(message, true);
    return hex(privateKey.sign(digest, TWCurveStarkex));
}

bool MessageSigner::verifyMessage(const PublicKey& publicKey, const std::string& message, const std::string& signature) noexcept {
    auto starkSignature = parse_hex(signature, true);
    auto digest = parse_hex(message, true);
    return publicKey.verify(starkSignature, digest);
}

} // namespace TW::StarkEx

File: rust/tw_any_coin/src/message_signer.rs
// SPDX-License-Identifier: Apache-2.0
//
// Copyright © 2017 Trust Wallet.

use tw_coin_entry::error::SigningResult;
use tw_coin_registry::coin_type::CoinType;
use tw_coin_registry::dispatcher::coin_dispatcher;
use tw_memory::Data;

/// Represents a message signer to sign regular or typed structured data for any blockchain.
pub struct MessageSigner;

impl MessageSigner {
    /// Signs a message.
    #[inline]
    pub fn sign_message(input: &[u8], coin: CoinType) -> SigningResult<Data> {
        let (ctx, entry) = coin_dispatcher(coin)?;
        entry.sign_message(&ctx, input)
    }

    /// Computes preimage hashes of a message.
    #[inline]
    pub fn message_preimage_hashes(input: &[u8], coin: CoinType) -> SigningResult<Data> {
        let (ctx, entry) = coin_dispatcher(coin)?;
        entry.message_preimage_hashes(&ctx, input)
    }

    /// Verifies a signature for a message.
    #[inline]
    pub fn verify_message(input: &[u8], coin: CoinType) -> SigningResult<bool> {
        let (ctx, entry) = coin_dispatcher(coin)?;
        entry.verify_message(&ctx, input)
    }
}
